com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace

```diff
@@ -66,12 +77,46 @@ void RayGen()
     jitteredPixelCoord.x += GetSample(currentPixelCoord, _RaytracingSampleIndex, 40);
     jitteredPixelCoord.y += GetSample(currentPixelCoord, _RaytracingSampleIndex, 41);
 
-    // Compute the ray direction, from those coordinates
-    float3 directionWS = -normalize(mul(jitteredPixelCoord, (float3x3)_PixelCoordToViewDirWS));
+    float3 directionWS;
+    float3 cameraPosWS;
+
+    float apertureRadius = _PathTracedDoFConstants.x;
+    if (apertureRadius == 0.0)
+    {
+        // Compute the ray direction from those coordinates (fast path for zero aperture)
+        directionWS = -normalize(mul(jitteredPixelCoord, (float3x3)_PixelCoordToViewDirWS));
+        cameraPosWS = _WorldSpaceCameraPos;
+    }
+    else
+    {
+        // Compute the ray origin and direction for a lens with non-zero aperture
+
+        // Apply the inverse viewport transform to go from vieport coordinates to NDC
+        jitteredPixelCoord = mul(jitteredPixelCoord, (float3x3)_InvViewportTransform);
+
+        // Sample the lens apperture
+        float r1 = GetSample(currentPixelCoord, _RaytracingSampleIndex, 106);
+        float r2 = GetSample(currentPixelCoord, _RaytracingSampleIndex, 107);
+        float2 uv = UniformPointWithinCircle(apertureRadius, r1, r2);
+
+        // Compute the new ray origin ( _ViewMatrix[0] = right, _ViewMatrix[1] = up, _ViewMatrix[2] = forward )
+        float focusDistance = _PathTracedDoFConstants.y;
+        float3 focusPoint = _WorldSpaceCameraPos - _ViewMatrix[2] * focusDistance;
+        cameraPosWS = _WorldSpaceCameraPos + _ViewMatrix[0] * uv.x + _ViewMatrix[1] * uv.y;
+
+        // Create the new orthonormal basis
+        float3 newForward = normalize(cameraPosWS - focusPoint);
```

Comment:
We have already function to create orthogonal basis, please use them :)

Reply:
I think GetLocalFrame() was the closest thing to what I need here. Not sure if there is something even better.